### PR TITLE
Feat/crd argo rollouts

### PR DESCRIPTION
**Pull request by oliveiraxavier**
This PR allow get some unused Argo rollouts

## PR Checklist

- [x] This PR adds K8s exceptions (false positives)
- [x] This PR adds new code
- [x] This PR includes tests for new/existing code
- [x] This PR adds docs

## Notes for your reviewers
This PR adds support for  get unused Argo Rollout, Cluster Analysis templates and  Analysis templates.

For this pr, some code refactoring was necessary.

### CODE REVIEW

**Comment by yonahd** [comment]
@oliveiraxavier Please rebase so we can properly review the code

**Comment by oliveiraxavier** [comment]
@yonahd  The main branch has been merged with the feat branch.

**Comment by yonahd** [comment]
> @yonahd  The main branch has been merged with the feat branch. 
> 

Did you sync your fork?

**Comment by oliveiraxavier** [comment]
> > @yonahd  The main branch has been merged with the feat branch.
> 
> Did you sync your fork?

Yes, synced

**Comment by yonahd** [comment]
@oliveiraxavier I'm working on the review. Any chance you can split this into 2 separate PRs
One for the Argo rollout and a second for moving the create client logic into a separate folder?

**Comment by oliveiraxavier** [comment]
> @oliveiraxavier I'm working on the review. Any chance you can split this into 2 separate PRs One for the Argo rollout and a second for moving the create client logic into a separate folder?


In my previous PR (https://github.com/yonahd/kor/pull/363), the logic was separated but the resource search files (deployments for example) were extremely tied to the connection logic, I separated them with that in mind. Refactoring this would almost be a reopening of the previous pr and would require some more refactoring of everything I developed, I can't see how I can split this PR into two without breaking the functioning of everything.

**Comment by yonahd** [comment]
> > @oliveiraxavier I'm working on the review. Any chance you can split this into 2 separate PRs One for the Argo rollout and a second for moving the create client logic into a separate folder?
> 
> In my previous PR (#363), the logic was separated but the resource search files (deployments for example) were extremely tied to the connection logic, I separated them with that in mind. Refactoring this would almost be a reopening of the previous pr and would require some more refactoring of everything I developed, I can't see how I can split this PR into two without breaking the functioning of everything.

Maybe I'm missing something but it seems most of the connection logic is just moving the logic from kor.go 
If it's too much to separate I'll review it as it is (mainly wanted the separation to help with my adhd)

**Comment by oliveiraxavier** [comment]
> Talvez eu esteja esquecendo de algo, mas parece que a maior parte da lógica de conexão está apenas movendo a lógica do kor.go. Se for muita coisa para separar, revisarei como está (queria principalmente a separação para ajudar com meu TDAH)

I won't be able to work on it at the moment

File: pkg/clusterconfig/connection.go
Comment (yonahd) [inline pkg/clusterconfig/connection.go:1]: Please move this into utils

File: pkg/kor/formatter.go [hunk 1]
```diff
@@ -235,3 +235,12 @@ func FormatOutputAll(namespace string, allDiffs []ResourceDiff, opts common.Opts
 	table.Render()
 	return fmt.Sprintf("Unused resources in namespace: %q\n%s\n", namespace, buf.String())
 }
+
+func SkipIfContainsValue(data []ResourceInfo, key string, value interface{}) bool {
```
Comment (yonahd) [inline pkg/kor/formatter.go:239]: This Should not be in formatter.go

File: pkg/kor/argorollouts.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	"fmt"
+	"os"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/yonahd/kor/pkg/clusterconfig"
+	"github.com/yonahd/kor/pkg/filters"
+)
+
+func GetUnusedArgoRollouts(clientsetinterface clusterconfig.ClientInterface, namespace string, filterOpts *filters.Options) ResourceDiff {
```
Comment (yonahd) [inline pkg/kor/argorollouts.go:14]: We need to create cmd/kor/argorollouts.go so users can select this one specifically 
Also missing pkg/kor/multi.go - allow finding your new resource in a comma-separated query along other resources.

File: pkg/kor/argorollouts.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+	"context"
+	"fmt"
+	"os"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/yonahd/kor/pkg/clusterconfig"
+	"github.com/yonahd/kor/pkg/filters"
+)
+
```
Comment (yonahd) [inline pkg/kor/argorollouts.go:13]: nit: move the get unused functions to the bottom to match convention

File: pkg/kor/config_test.go
Comment (yonahd) [inline pkg/kor/config_test.go:1]: What is this needed for?

**Review by yonahd — COMMENTED** [review]
Logic looks right for the finding unused argorollouts
There are some structure issues and minor fixes.
I suggest you also look through https://github.com/yonahd/kor/blob/main/CONTRIBUTING.md which will help structuring the PR.
Also we need to think of how we give the permissions in the helm chart

File: pkg/kor/argorollouts.go
[the remark this answers is not in the document]
Comment (oliveiraxavier) [reply pkg/kor/argorollouts.go:1]: This should cause circular import error

**Comment by oliveiraxavier** [comment]
I'm closing this pr. 
My time is short and until I finish these changes there will be many conflicts.

**Comment by oliveiraxavier** [comment]
Thank's for your review

**Comment by yonahd** [comment]
> I'm closing this pr. My time is short and until I finish these changes there will be many conflicts.

Are you planning on completing this?

**Comment by oliveiraxavier** [comment]
> > I'm closing this pr. My time is short and until I finish these changes there will be many conflicts.
> 
> Are you planning on completing this?

Maybe in two or three months I can set aside some time to get back to it. If someone follows the idea, no problem.

**Pull request closed without merging**